src/bridge/cockpitrouter.c

```diff
@@ -1713,73 +1711,43 @@ cockpit_router_dbus_startup (CockpitRouter *router)
 
 /* Superuser init */
 
-static void superuser_init_step (CockpitRouter *router);
-
-static void
-superuser_init_start (CockpitRouter *router,
-                      const gchar *id)
-{
-  router->superuser_init_in_progress = TRUE;
-  router->superuser_init_next_rule = router->rules;
-  router->superuser_init_id = g_strdup (id);
-
-  superuser_init_step (router);
-}
-
 static void
-superuser_init_step_done (const gchar *error, gpointer user_data)
+superuser_init_done (const gchar *error, gpointer user_data)
 {
   CockpitRouter *router = user_data;
 
   if (error)
-    {
-      router->superuser_rule = NULL;
-      superuser_init_step (router);
-    }
-  else
-    {
-      router->superuser_init_in_progress = FALSE;
-      g_free (router->superuser_init_id);
-      router->superuser_init_id = NULL;
-      superuser_notify_property (router, "Current");
+    router->superuser_rule = NULL;
 
-      if (!router->superuser_legacy_init)
-        {
-          GBytes *request = cockpit_transport_build_control ("command", "superuser-init-done",
-                                                             NULL);
-          cockpit_transport_send (router->transport, NULL, request);
-          g_bytes_unref (request);
-        }
+  router->superuser_init_in_progress = FALSE;
+  superuser_notify_property (router, "Current");
+
+  if (!router->superuser_legacy_init)
+    {
+      GBytes *request = cockpit_transport_build_control ("command", "superuser-init-done",
+                                                         NULL);
+      cockpit_transport_send (router->transport, NULL, request);
+      g_bytes_unref (request);
     }
 
   g_object_unref (router);
 }
 
 static void
-superuser_init_step (CockpitRouter *router)
+superuser_init_start (CockpitRouter *router,
+                      const gchar *id)
 {
-  for (GList *l = router->superuser_init_next_rule; l; l = g_list_next (l))
-    {
-      RouterRule *rule = l->data;
-      gchar *rule_id = rule_superuser_id (rule);
-      gboolean only_explicitly = FALSE;
-
-      // XXX - This is a hack to make sure that "any" will always
-      // choose "sudo".  The rest of the code is not really ready for
-      // anything else.  Once it is, there might not be any need for
-      // "only-explicitly" any more.
-
-      if (rule->config)
-        cockpit_json_get_bool (rule->config, "only-explicitly", FALSE, &only_explicitly);
+  router->superuser_init_in_progress = TRUE;
 
-      if (rule_id && ((router->superuser_init_id == NULL && !only_explicitly)
-                      || (router->superuser_init_id != NULL && g_str_equal (router->superuser_init_id, rule_id))))
+  for (GList *l = router->rules; l; l = g_list_next (l))
+    {
+      gchar *rule_id = rule_superuser_id (l->data);
+      if (rule_id && (id == NULL || g_str_equal (id, rule_id)))
```

Comment:
so "any" means `id == NULL`, right? That seems fine. But to be on the safe side, should we then move the sudo rule in the manifest to be the first one, to be on the safe side?

Reply:
It is the first, no?

Reply:
> so "any" means `id == NULL`, right?

Yes, exactly.

Reply:
Argh, sorry -- I was looking at my test branch, which had it in the other order.